### ISSUE TITLE
Provide a bash compatible syntax for bats to allow for other tools to work

### ISSUE THREAD

**Issue by lucastheisen** [bats-core/bats-core#192]
Right now, the `.bats` files are not valid shell syntax so external tools, [like shellcheck, fail when they try to parse the script](https://github.com/koalaman/shellcheck/issues/709).  This makes it difficult to leverage the benefits of static analysis, while still using this library.  We tried a few workarounds, but all had problems.  It is a simple matter [to support an _additional_ syntax](https://github.com/bats-core/bats-core/pull/191) which is valid bash, allowing for shellcheck to succeed.

**Comment by sublimino** [comment]
Some thoughts:
- are there other syntax types that are more preferable? Options include function names pre/suffixed with strings such as `test_`, annotations/comments, or other magic strings
- could treating comments as code break existing well-known Bats test suites (investigation required)
- what is the rationale for the existing syntax (anything other than parsability?)
- are other linters a target (e.g. shfmt)?

Thanks for starting the conversation @lucastheisen, I agree existing syntax isn't compatible with established tooling. This is a significant enough change that it may warrant a major release anyway. Thoughts welcome!

**Comment by waterkip** [comment]
With Perl one has [Test::Class::Moose](https://metacpan.org/pod/Test::Class::Moose#Declare-a-test-method) where you can have `sub test_foo { ... }` or `sub foo : Test { ... }` or `sub foo : Tests { ... }`. Support for all three of them is very nice. 

I'm in favor of this change, my linter also complains within vim with .bats files.

**Comment by ko1nksm** [comment]
How about this? It is similar to the current syntax. (Add a semicolon after the description)

```sh
@test "addition using bc"; {
  result="$(echo 2+2 | bc)"
  [ "$result" -eq 4 ]
}
```

I think also good to be able to use new line instead of a semicolon.

```sh
@test "addition using bc"
{
  result="$(echo 2+2 | bc)"
  [ "$result" -eq 4 ]
}
```

**Comment by spl** [comment]
I found this issue after encountering the error below when running [`shfmt`](https://github.com/mvdan/sh) on a `.bats` file.

```
test.bats:23:1: "}" can only be used to close a block
```

Are there any known workarounds for using `shfmt` on Bats files?

**Comment by jasonkarns** [comment]
The bats preprocessor generates valid bash. If we support an additional syntax that is valid bash from the get-go, it seems that using the same syntax bats itself generates would be preferable?

**Comment by spl** [comment]
For anyone who might be interested, this is what I did.

All my Bats tests are found in `./test/*.bats` and written with this pattern:

```sh
@test "test description" {
  # ...
}
```

If I change the above by inserted a newline before the `{`, then `shfmt` has no problem with it. The result would be:

```sh
@test "test description"
{
  # ...
}
```

The following script uses in-place edits to implement this change before `shfmt` and reverse the same change after `shfmt`:

```sh
# Rewrite the Bats scripts in-place to look more like Bash scripts to shfmt
perl -pi -e 's/^(\@test.*) \{$/$1\n{/' ./tests/*.bats

# Run on all Bats scripts
shfmt ./tests/*.bats

# Undo the changes to the Bats scripts in-place so that they work with Bats
perl -pi -e 's/^\{\R//; s/(\@test.*$)/$1 {/' ./tests/*.bats
```

**Comment by virgilwashere** [comment]
> using the same syntax bats itself generates would be preferable?
Sounds good to me, unless said syntax is human unfriendly??

Easy example ?

**Comment by lucastheisen** [comment]
@sublimino , your [original comment](https://github.com/bats-core/bats-core/issues/192#issuecomment-468028089) indicated more evaluation was necessary.  Has that happened.  As stated in [this comment](https://github.com/bats-core/bats-core/pull/191#issuecomment-467980185) my pull request 

> was designed to avoid
>
> >   a breaking change

It makes for a _pure bash_ syntax with no other side effects and a very small modification.  I have been working off of that branch for almost a year now (with no issues), but it would be nice to get back to the mainline and benefit from all the other fixes applied to the project without re-merging back into my fork...

**Comment by jersou** [comment]
a dirty workaround to have the syntax
```
@test "test description"
{
  ...
}
```
in [libexec/bats-core/bats-preprocess line 54](https://github.com/bats-core/bats-core/blob/master/libexec/bats-core/bats-preprocess#L54):
```
- } <<< "$(< "$test_file")"$'\n'
+ } <<< "$(sed -z 's|\n{\n| {\n|g' "$test_file")"$'\n'
```

**Comment by marwatk** [comment]
I've been using @lucastheisen's [fork](https://github.com/pastdev/bats-core) which works great. Is there anything specific holding up getting proper bash syntax on the tests? Seems like his is a non-breaking change with an alternative syntax that won't accidentally include functions that weren't meant to. (Whereas adding a prefix or postfix like `test_` or `_test` could accidentally include functions that weren't intended to be real tests).

**Comment by lucastheisen** [comment]
Just merged the latest updates into [my fork/PR](https://github.com/bats-core/bats-core/pull/191) so there are again no merge conflicts.  Is there anything else we can do to help get this patch merged?

**Comment by marwatk** [comment]
I tried to test various linters on the semicolon version suggested [here](https://github.com/bats-core/bats-core/issues/192#issuecomment-512282673) (which probably would have been cleaner to implement as it could keep a single regex) but there are two issues with it. First, `shfmt` won't preserve it:

```bash
$ cat test/fixtures/bats/semicolon_style.bats
@test "should be found"; {
  true
}
mwatkins@MM233580-PC ~/tmp/marwatk-bats-core [master ?]
$ ./shfmt_v3.1.0_linux_amd64 test/fixtures/bats/semicolon_style.bats
@test "should be found"
{
  true
}
```
But the larger issue is that `shellcheck` now supports bats natively, and the `;` format breaks it:
```bash
In test/fixtures/bats/semicolon_style.bats line 1:
@test "should be found"; {
^-- SC1073: Couldn't parse this bats @test. Fix to allow more checks.
                       ^-- SC1072:  Fix any mentioned problems and try again.
```

So previously the `;` solution could have worked but now it's a breaking change.

The comment version from @lucastheisen survives `shfmt` and still works with `shellcheck` so I don't see another way forward that isn't a breaking change. It's also most likely to work with the widest variety of other tooling as it's unlikely the tooling would move the comment without _also_ breaking bats by moving the opening `{`. 

So the only downside I see is that it treats comments as code. And it's just an _alternative_ syntax, the old one continues to work.

**Comment by lucastheisen** [comment]
Is there some discussion still needed on this topic?  This issue, and [the associated PR](https://github.com/bats-core/bats-core/pull/191) have been around for a year and a half.  Its a very _simple_ non-intrusive, non-required, syntax option that just works.  The maintenance load should be extremely small (basically a single regex).

If there is an argument against this perhaps we could have that discussion here and close this issue out if its never going to get merged.  Either way, it would be nice to know...

**Comment by sublimino** [comment]
Thanks for all the thought that has gone into this @lucastheisen and others (and for chasing!), in princple I'm happy with this change. Anything further to add here @jasonkarns @martin-schulze-vireso @waterkip @dimo414 @ko1nksm? Or anybody else?

Assuming no objections I'll look to merge https://github.com/bats-core/bats-core/pull/191 in the next couple of weeks.

**Comment by martin-schulze-vireso** [comment]
I somewhat dislike the loss of descriptivity by having to name tests as bash functions with this. Since this runs through preprocessor anyways, why not reuse the same function name again and again:

```
function bats_test() { # name: Some descriptive title for this test
  ...
}

function bats_test() { # name: Some other test
  ....
}
```

This should still be valid bash while retaining the free form naming.

**Comment by lucastheisen** [comment]
> I somewhat dislike the loss of descriptivity by having to name tests as bash functions with this

I have been using this fork for a couple years now and do not find a loss of descriptivity.  Instead of spaces we just use underscores:

```bash
function some_descriptive_title_for_this_test { # @test
}

function some_other_test { # @test
}
```

I did think about other approaches like you mention but didn't find value.  With that approach you still need a unique function name, so in practice it would usually end up simply duplicating:

```bash
function some_descriptive_title_for_this_test { # @name: Some descriptive title for this test
}
```

Then this introduces confusion as to _which_ name any `--filter` would match.  Would it be the descriptive title?  Or the function name?

> we should go all in and opt for a test_ prefix. 

I don't disagree with any of this, but don't think it should preclude [this pull request|https://github.com/bats-core/bats-core/pull/191] in any way.  I think the performance improvement and simplicity of removing the pre-processor would be ideal, but this PR works now, with full backwards compatibility.  The only cost is relatively minimal in a second regex test of the very short strings.

**Comment by martin-schulze-vireso** [comment]
Just let me state upfront that I do believe this code is ready to be merged. I am just late for the discussion since I first saw this issue through @sublimino's mention and wanted to add my 2ct with a new longterm perspective.

> couple years now and do not find a loss of descriptivity.

I am not strongly opposed to that, having to conform to bash names just reduces the expressivity somewhat but many testing frameworks like junit already mandate that anyways.

> With that approach you still need a unique function name

No, as we still use the preprocessor, the comment part can be used to generate a unique function name like it is now with the string after `@test". In my proposal, the function name is only there to please bash/shellcheck and can be used as  a test function marker as opposed to helper functions, e.g.:

```
function bats_test { # this is the real name of the test, note the missing `@test` or `@name`, which is not needed anymore
}

function helper_function { # this won't be run as a bats test
}
```

Neither bash nor shellcheck seem to be disturbed by the function being (re-)declared multiple times. In essence my proposal is to replace `@test` by (`function`) `bats_test` and move the test name into the comment at the end of the line, thus gaining fully bash compliant code while retaining all of the possible test name strings.

I believe the main issue at stake here is: 

> I have been using this fork for a couple years now 

Your implementation has been in use for a while now and you (and others) have an existing codebase that you understandably would not want to change.

My points are coming from another perspective: If we would start out without an existing user base for the this PR's implementation I don't see an advantage over my proposal.

Given the already existing codebase I would vote for your code to go into the current major version. However, for a new major release that might not need to maintain backwards compatibility, I would like to consider the other approaches, especially the one without preprocessor.

**Comment by lucastheisen** [comment]
> No, as we still use the preprocessor, the comment part can be used to generate a unique function name like it is now with the string after @test. In my proposal, the function name is only there to please bash/shellcheck and can be used as a test function marker as opposed to helper functions
> ...
> shellcheck does not seem to be disturbed by the function bats_test being (re-)declared multiple times

I hear what you are saying.  The only _problems_ i could think of are:

- might be confusing to people reading the test file to see function named the same
- while valid bash, if run as bash (ie: sourced inside another file), each test would replace the previous which is almost certainly not desired

> My points are coming from another perspective: If we would start out without an existing user base for the this PR's implementation I don't see an advantage over my proposal.

From that perspective, i mostly agree.  If i were to design from the ground up i would start with requiring valid bash, but from there, i see the forced `test_` prefix as one of the 2 most common approaches, the other being annotation (java)/attribute (c#).  Given that bash has no formal _annotation_ the conventional `test_` prefix is likely the most natural (with the drawback being that you couldn't have any helper functions whose names start with `test_`.  For example, you may have:

```
# not a test, just a utility that tests the supplied data for some condition
function test_data {
  local data=$1
  ...
}

function should_be_valid { #23 
  if ! test_data "$(get_valid_data)"; then ext 1; fi
}

function should_be_invalid { #@test
  if test_data "$(get_invalid_data)"; then exit 1; fi
}
```

Of course i could just rename `test_data` to something like `assert_data` and its no longer an issue, but its a _special_ case and those eventually have issues (for someone).

Using comment style annotations, that is not a problem.  Something like:

```bash
# @test @condition(Should be invalid)
function any_name {
}
```

But now you've created a new syntax you have to parse, comments mean something other than commentary, and it would be far less efficient than your suggested approach for `test_` prefix.

In summation,

> for a new major release that might not need to maintain backwards compatibility

I think the `test_` prefix is probably the best way to go!

**Comment by Mayeu** [comment]
This may be of interest of follower of this issue: both shellcheck and shfmt added support for bats file:
- [shellcheck 0.7.0](https://github.com/koalaman/shellcheck/blob/master/CHANGELOG.md#v070---2019-07-28)
- [shfmt 3.2.0](https://github.com/mvdan/sh/releases/tag/v3.2.0)

**Comment by martin-schulze-vireso** [comment]
Since this has been largely solved with #378 I am closing this in favor of #306. Once that is solved, we will have a trivial path towards implementing function name base auto registration. (So much so, that it could be implemented by users themselves.)
